### PR TITLE
fix DS-13, DS-22, other changes

### PR DESCRIPTION
**Pull request by ItsNotJiaching**
The bot now scrapes uploaded information as an .md file directly to the GitHub wiki repository, and no longer puts all of the info in a markdown file called “messages.md”.

1. added the subprocess module, allowing us to run terminal commands.
2. after the third question is answered, the bot will commit & push the new wiki file to the wiki repo.
3. each time the wiki repo were to be read/updated, the bot first pulls from the cloud before reading/changing anything.
4. replaced all calls to [messages.md](http://messages.md) and replaced with appropriate [docs-slackbot.wiki](http://docs-slackbot.wiki) calls. This way, the bot will both CHECK whether a timestamp is scraped in the github repo AND WRITE a new page in the local clone of the wiki repo.

other changes:

1. added a function that checks whether an exact title has been used before. this is incomplete — refer to code comments and DS-25 on Jira.
2. changed description of the AttributeError to reflect new findings
3. changed organization of how content is scraped to make it look cleaner.
4. replaced 3 bot questions variables with one array in bot_messages.py
5. updated comments to explain new code
6. added note to onboarding message that pins don’t activate the bot

### CODE REVIEW

**Comment by ItsNotJiaching** [comment]
question & comment:

question - I used to use global variables a lot when I was younger & I've been told they're pretty bad in most cases. I wonder if the usage of wiki_title warrants it being a global variable. Should it be a global title, or is there a better way to do the same thing with a local variable?

comment - I know you showed me how to check whether a file exists using the command line (subprocess), but I resorted to what I felt comfortable with and used os.exists() to check whether a file exists. Don't know if that may be a future issue.

File: app.py [hunk 1]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
                 # compares the thread message history to the individual questions. 
                 # if a the exact string of a question is in message["text"], change the value of last_question_asked accordingly.
                 for message in thread_message_history:
-                    if message["text"] == bot_messages.question_one:
+                    if message["text"] == bot_messages.questions[0]:
                         last_question_asked = 1
```
Comment (iengel35) [inline app.py:137]: Nit: if you start `last_question_asked` at 0, you can use it to index your `questions` array.

File: app.py [hunk 2]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
...
-                    client.chat_postMessage(channel=channel_id, text=bot_messages.response_recorded, thread_ts=message_thread_ts)
-                    client.chat_postMessage(channel=channel_id, text=bot_messages.question_two, thread_ts=message_ts)
-                    file_write(f"# {text}\n")
-                # bot records response to question 2 and asks question 3 afterwards.
+                    if not title_in_documentation(text):
+                        client.chat_postMessage(channel=channel_id, text=bot_messages.response_recorded, thread_ts=message_thread_ts)
+                        client.chat_postMessage(channel=channel_id, text=bot_messages.questions[1], thread_ts=message_ts)
+                        wiki_title = text
+                    else:
+                        client.chat_postMessage(channel=channel_id, text=bot_messages.already_titled, thread_ts=message_ts)
```
Comment (iengel35) [inline app.py:151]: A possible future addition: if the title already exists, you can link the existing entry and/or ask if they would like to append the pinned message to the existing page.

File: app.py [hunk 3]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
...
+                # bot writes response to question 2 into markdown file and asks question 3 afterwards.
```
Comment (iengel35) [inline app.py]: Nit: would be helpful to add a brief comment about what each question is asking for right before you post it (i.e. question one asking for title, 2 for picture, 3 for supplemental information)

File: app.py [hunk 4]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
...
                     client.chat_postMessage(channel=channel_id, text=bot_messages.wiki_page_made, thread_ts=message_ts)
                     scraped_message = thread_message_history[0]["text"]
                     if text == "None":
-                        file_write(f"## Original Documentation:\n\n{scraped_message}\n\n## Additional Info:\n\n###### Message Timestamp: {message_thread_ts}\n\n")
+                        file_write(f"## Original Documentation\n\n{scraped_message}\n\n## Additional Info\n\nMessage Timestamp: {message_thread_ts}\n\n", wiki_title)
                     else:
-                        file_write(f"## Original Documentation:\n\n{scraped_message}\n\n## Additional Info:\n\n{text}\n\n###### Message Timestamp: {message_thread_ts}\n\n")
+                        file_write(f"## Original Documentation\n\n{scraped_message}\n\n## Additional Info\n\n{text}\n\nMessage Timestamp: {message_thread_ts}\n\n", wiki_title)
+                    subprocess.call(['git', 'add', f"{wiki_title}.md"], cwd="../docs-slackbot.wiki")
+                    subprocess.call(['git', 'commit', '-am', f"A new wiki named {wiki_title} has been uploaded."], cwd="../docs-slackbot.wiki")
```
Comment (iengel35) [inline app.py:174]: Nit: commit message should include that the wiki page has been uploaded by the slackbot.

**Review by iengel35 — APPROVED** [review]
Made a couple suggestions, but nothing blocking for you to merge into main.

File: app.py [hunk 5]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
...
+                    subprocess.call(['git', 'push', '-f'], cwd="../docs-slackbot.wiki")
 
     # ========================= Code End ========================= #
```
Comment (iengel35) [inline app.py]: Why is this comment here?

File: app.py [hunk 6]
```diff
@@ -127,53 +133,82 @@ def message(event, client):
...
+#     ack()
+#     try:
+#         if command["text"].lower() == "on":
+#             respond("Other users can now scrape your messages.")
+#         elif command["text"].lower() == "off":
+#             respond("Other users will now be unable to scrape your messages.")
+#     # This is a KeyError. It occurs when the user did not input text alongside the command.
+#     # In other words, command["text"] returned no text.
+#     except KeyError:
+#         return respond("Please type on or off to change whether you want others to scrape your messages.")
```
Comment (iengel35) [inline app.py]: Try to avoid including commented out code in a PR

**Comment by iengel35** [comment]
And to reply to your questions- 
1 -  yes, global variables should be avoided where possible because A they mean that any function can modify this data which is relied on by other functions, so it will not always contain the value a function expects it to if it has been modified by some other function. And B, they increase the complexity of your code- you can't just go over a single function because you need to look at the entire file to understand what a global variable will hold.
In this case, I think it is fine to keep `wiki_title` global. Another option would be to create a python bot class which could then store wiki title as member data, so it would persist between function calls. But your current solution works, so it's ok for now!

2 - os.exists is good! Probably the more direct way to check for the file than my suggestion.

File: app.py (hunk 5 above)
Comment (ItsNotJiaching) [reply app.py]: this has been here for a while -- I used it to pretty much block out where the bulk of the code for the bot will be, but now I don't think it necessarily contributes to code clarity.

File: app.py (hunk 4 above)
Comment (ItsNotJiaching) [reply app.py:174]: will make the small change! i didn't think about that.

File: app.py (hunk 3 above)
Comment (ItsNotJiaching) [reply app.py]: I was debating between writing either just "question 3" or what the question asked for, but now that you point it out, the latter definitely makes more sense.

File: app.py (hunk 2 above)
Comment (ItsNotJiaching) [reply app.py:151]: that was the plan, but given limited time I had to compromise. 

in addition, i did forget to put it on the jira backlog, so i just added that!

File: app.py (hunk 6 above)
Comment (ItsNotJiaching) [reply app.py]: fixed -- i kept my commented code somewhere else.

**Pull request merged**